Don't run user function until all UserRRefs in the args are confirmed

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34497 Don't run user function until all UserRRefs in the args are confirmed**
* #34809 Fix Error Hanlding for Local Send
* #34496 Split deserialize from runPythonUdf and remove generatePythonUDFResult

Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

mrshenli committed:
[WIP] Don't run user function until all UserRRefs in the args are confirmed

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Don't run user function until all UserRRefs in the args are confirmed"




[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Don't run user function until all UserRRefs in the args are confirmed"




[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Don't run user function until all UserRRefs in the args are confirmed"




[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Don't run user function until all UserRRefs in the args are confirmed"




[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/rref_impl.h:
```diff
@@ -290,6 +299,10 @@ class TORCH_API OwnerRRef final : public RRef {
     return true;
   }
 
+  inline bool isConfirmed() const override {
```
Should we add a comment here that for OwnerRRef it's always confirmed to avoid confusion?

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1651,3 +1654,33 @@ def test_function_not_on_callee(self):
+        self.assertEqual(ret, True)
+
+    @dist_init
+    def test_user_rrefs_confirmed_remote(self):
```
Is it possible to add a test case where a UserRRef is never confirmed and verify the appropriate behavior there? (i.e. stall forever and not run the function, throw an appropriate exception, etc).

mrshenli replied:
I thought about this. Looks like we would need to deplete the threads on the user side or owner side to prevent them either from sending out the RRef request or confirmation. But that threading behavior will change when we introduce async user function. Let me spend some more time to see if it is possible to test it without messing up with threads.

rohan-varma commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -386,9 +406,26 @@ void RRefContext::delPendingUser(const ForkId& forkId) {
+    state->future_.wait();
+  }
+  userTable_.clear();
+  recording = false;
```
Why do we want to stop recording additional pending users once we've called `waitForThreadLocalPendingUsers`?

mrshenli replied:
pending users are added into the context for two reasons:

1. when we call `rpc.remote`, the return value `UserRRef` on the caller will be added to the context.
2. when we forward an RRef to another user, the `UserRRef` on the destination will be added to the context. 

For 1, we don't block as we would like to use that `UserRRef` right away. We only block in 2. So, if we don't close the recording session, there could be more `UserRRef`s added into the `userTable_`. Consider a case where the application runs a user function remotely, and the user function creates a RRef using `rpc.remote`. This RRef should not block the next user function handled by the same RPC thread. 

Thanks for pointing this out, let me add this description to the code.

rohan-varma reviewed (approved):
Looks good, some minor questions. Although we should also get a review from @pritamdamania87 or @zhaojuanmao.

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -147,6 +148,14 @@ class TORCH_API RRefContext {
+  // Wait until all pending UserRRefs in userTable_ are confirmed by their
+  // owners and then clear the userTable_. This is invoked to make sure RRefs
+  // in user function args are confirmed before launching user code.
+  void waitForThreadLocalPendingUsers();
```
nit: PendingRRefs instead of PendingUsers

pritamdamania87 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -369,6 +379,9 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
 std::shared_ptr<FutureMessage> RequestCallbackImpl::processMessage(
     Message& request) const {
+  if (request.hasUserFunction()) {
+    RRefContext::getInstance().recordThreadLocalPendingUsers();
```
Shouldn't this apply to all Messages? If in the future I build a custom message for internal system messages and send an RRef in it, we should still honor the same protocol where we wait for confirmation right?

mrshenli replied:
Sure, will make the change.

pritamdamania87 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -50,6 +50,7 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
 
       // scriptCall is only alive within this block, use reference to avoid copy
       auto& stack = scriptCall.stackRef();
+      RRefContext::getInstance().waitForThreadLocalPendingUsers();
```
Should we just do this for all message types and also record for all message types?

mrshenli replied:
You mean this this line of code will appear in every switch-case clause? This cannot directly wrap the `deserializeRequest` function, because Python deserialization occurs within the clause. We probably can add a message layer on top of PythonCall to hold the deserialized py::object. Let me try.

pritamdamania87 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -107,6 +112,7 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
       // scriptRemoteCall is only alive within this block, use reference to
       // avoid copy
       auto& stack = scriptRemoteCall.stackRef();
+      RRefContext::getInstance().waitForThreadLocalPendingUsers();
```
Can we do this in an async way where we don't block the RPC thread? We can add callbacks on the futures in userTable_ to mark an 'uber' future as complete once done. Then, the 'uber' future can have a callback that actually does all the RPC execution itself. Once the RPC execution is done, we can mark the future that was returned to the RPC layer as complete so that it can respond back to the user.

Given we plan to merge RPC and JIT pools, we can't have blocking calls like these.

mrshenli replied:
Yes, this should work. let me try.

pritamdamania87 reviewed (changes requested):
Requesting changes since we should avoid blocking the RPC thread waiting for an RRef to be confirmed.

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

mrshenli committed:
Update on "Don't run user function until all UserRRefs in the args are confirmed"


Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -379,11 +450,42 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
+        });
+  } catch (std::exception& e) {
+    retFuture->markCompleted(handleError(e, request.type(), request.id()));
+    rrefContext.clearRecordedPendingRRefsOnError();
```
nit: better to do this with RAII.

mrshenli replied:
Will do this in the next PR
